File: crates/traderview-core/src/standard_vs_itemized.rs
//! Standard vs itemized deduction — whether a taxpayer is better off taking the
//! standard deduction or itemizing on Schedule A.
//!
//! Itemized total = SALT (state/local income or sales tax + property tax, capped)
//! + home mortgage interest + charitable contributions + medical expenses above
//! the 7.5%-of-AGI floor. The taxpayer takes the larger of that and the standard
//! deduction.
//!
//! 2026 defaults (web-verified, OBBBA, overridable): standard deduction $16,100
//! single / $32,200 MFJ; SALT cap $40,400, reduced 30% of MAGI over $505,000 and
//! floored at $10,000 (the cap reverts to $10,000 in 2030); medical floor 7.5%.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FilingStatus {
    Single,
    MarriedJoint,
}

fn d_salt_cap() -> f64 {
    40_400.0
}
fn d_salt_threshold() -> f64 {
    505_000.0
}
fn d_salt_floor() -> f64 {
    10_000.0
}
fn d_medical_floor_pct() -> f64 {
    7.5
}

#[derive(Debug, Clone, Deserialize)]
pub struct StdVsItemizedInput {
    /// Adjusted gross income (used as MAGI for the SALT phase-out and the
    /// medical floor).
    pub agi_usd: f64,
    pub filing_status: FilingStatus,
    /// State/local income or sales tax paid.
    #[serde(default)]
    pub state_local_tax_usd: f64,
    #[serde(default)]
    pub property_tax_usd: f64,
    #[serde(default)]
    pub mortgage_interest_usd: f64,
    #[serde(default)]
    pub charitable_usd: f64,
    #[serde(default)]
    pub medical_usd: f64,
    /// Base SALT cap before the high-income phase-out.
    #[serde(default = "d_salt_cap")]
    pub salt_cap_base_usd: f64,
    /// MAGI where the SALT cap begins to phase down.
    #[serde(default = "d_salt_threshold")]
    pub salt_phaseout_threshold_usd: f64,
    /// SALT cap floor the phase-out cannot drop below.
    #[serde(default = "d_salt_floor")]
    pub salt_floor_usd: f64,
    /// Medical expense floor, percent of AGI.
    #[serde(default = "d_medical_floor_pct")]
    pub medical_floor_pct: f64,
    /// Standard deduction. None → filing-status default.
    #[serde(default)]
    pub std_deduction_usd: Option<f64>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct StdVsItemizedResult {
    pub standard_deduction_usd: f64,
    pub salt_paid_usd: f64,
    /// SALT cap after the high-income phase-out.
    pub salt_cap_usd: f64,
    pub salt_deductible_usd: f64,
    /// SALT paid above the cap (lost).
    pub salt_lost_usd: f64,
    pub medical_floor_usd: f64,
    pub medical_deductible_usd: f64,
    pub itemized_total_usd: f64,
    pub deduction_taken_usd: f64,
    /// Itemized − standard. Positive means itemizing wins.
    pub itemizing_advantage_usd: f64,
    pub should_itemize: bool,
}

pub fn analyze(input: &StdVsItemizedInput) -> StdVsItemizedResult {
    let standard = input.std_deduction_usd.unwrap_or(match input.filing_status {
        FilingStatus::Single => 16_100.0,
        FilingStatus::MarriedJoint => 32_200.0,
    });

    let salt_paid = input.state_local_tax_usd + input.property_tax_usd;
    let over = (input.agi_usd - input.salt_phaseout_threshold_usd).max(0.0);
    let salt_cap = (input.salt_cap_base_usd - 0.30 * over).max(input.salt_floor_usd);
    let salt_deductible = salt_paid.min(salt_cap);
    let salt_lost = (salt_paid - salt_deductible).max(0.0);

    let medical_floor = input.medical_floor_pct / 100.0 * input.agi_usd;
    let medical_deductible = (input.medical_usd - medical_floor).max(0.0);

    let itemized = salt_deductible
        + input.mortgage_interest_usd
        + input.charitable_usd
        + medical_deductible;

    let should_itemize = itemized > standard;

    StdVsItemizedResult {
        standard_deduction_usd: standard,
        salt_paid_usd: salt_paid,
        salt_cap_usd: salt_cap,
        salt_deductible_usd: salt_deductible,
        salt_lost_usd: salt_lost,
        medical_floor_usd: medical_floor,
        medical_deductible_usd: medical_deductible,
        itemized_total_usd: itemized,
        deduction_taken_usd: itemized.max(standard),
        itemizing_advantage_usd: itemized - standard,
        should_itemize,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-2
    }

    fn base() -> StdVsItemizedInput {
        StdVsItemizedInput {
            agi_usd: 100_000.0,
            filing_status: FilingStatus::Single,
            state_local_tax_usd: 0.0,
            property_tax_usd: 0.0,
            mortgage_interest_usd: 0.0,
            charitable_usd: 0.0,
            medical_usd: 0.0,
            salt_cap_base_usd: 40_400.0,
            salt_phaseout_threshold_usd: 505_000.0,
            salt_floor_usd: 10_000.0,
            medical_floor_pct: 7.5,
            std_deduction_usd: None,
        }
    }

    #[test]
    fn standard_wins_when_itemized_low() {
        let r = analyze(&StdVsItemizedInput {
            state_local_tax_usd: 5_000.0,
            mortgage_interest_usd: 3_000.0,
            charitable_usd: 1_000.0,
            ..base()
        });
        // itemized 9,000 < standard 16,100.
        assert!(close(r.itemized_total_usd, 9_000.0));
        assert!(close(r.standard_deduction_usd, 16_100.0));
        assert!(close(r.deduction_taken_usd, 16_100.0));
        assert!(!r.should_itemize);
        assert!(close(r.itemizing_advantage_usd, -7_100.0));
    }

    #[test]
    fn itemize_wins_for_mfj() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 200_000.0,
            filing_status: FilingStatus::MarriedJoint,
            state_local_tax_usd: 30_000.0,
            mortgage_interest_usd: 18_000.0,
            charitable_usd: 5_000.0,
            ..base()
        });
        // itemized 53,000 > standard 32,200.
        assert!(close(r.itemized_total_usd, 53_000.0));
        assert!(close(r.itemizing_advantage_usd, 20_800.0));
        assert!(r.should_itemize);
    }

    #[test]
    fn salt_cap_binds() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 300_000.0,
            filing_status: FilingStatus::MarriedJoint,
            state_local_tax_usd: 25_000.0,
            property_tax_usd: 20_000.0,
            ..base()
        });
        // paid 45,000; cap 40,400 (no phase-out); lost 4,600.
        assert!(close(r.salt_paid_usd, 45_000.0));
        assert!(close(r.salt_cap_usd, 40_400.0));
        assert!(close(r.salt_deductible_usd, 40_400.0));
        assert!(close(r.salt_lost_usd, 4_600.0));
    }

    #[test]
    fn salt_phaseout_partial() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 605_000.0,
            filing_status: FilingStatus::MarriedJoint,
            state_local_tax_usd: 50_000.0,
            ..base()
        });
        // 40,400 − 0.30 × (605,000 − 505,000) = 40,400 − 30,000 = 10,400.
        assert!(close(r.salt_cap_usd, 10_400.0));
        assert!(close(r.salt_deductible_usd, 10_400.0));
    }

    #[test]
    fn salt_phaseout_hits_floor() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 1_000_000.0,
            filing_status: FilingStatus::MarriedJoint,
            state_local_tax_usd: 50_000.0,
            ..base()
        });
        // 40,400 − 0.30 × 495,000 is negative → floored at 10,000.
        assert!(close(r.salt_cap_usd, 10_000.0));
        assert!(close(r.salt_deductible_usd, 10_000.0));
    }

    #[test]
    fn medical_above_floor() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 80_000.0,
            medical_usd: 10_000.0,
            ..base()
        });
        // floor 7.5% × 80,000 = 6,000; deductible 4,000.
        assert!(close(r.medical_floor_usd, 6_000.0));
        assert!(close(r.medical_deductible_usd, 4_000.0));
    }

    #[test]
    fn medical_below_floor_zero() {
        let r = analyze(&StdVsItemizedInput {
            agi_usd: 100_000.0,
            medical_usd: 5_000.0,
            ..base()
        });
        // floor 7,500 > 5,000 → no deduction.
        assert!(close(r.medical_deductible_usd, 0.0));
    }

    #[test]
    fn standard_deduction_override() {
        let r = analyze(&StdVsItemizedInput {
            charitable_usd: 5_000.0,
            std_deduction_usd: Some(30_000.0),
            ..base()
        });
        assert!(close(r.standard_deduction_usd, 30_000.0));
        assert!(close(r.deduction_taken_usd, 30_000.0));
        assert!(!r.should_itemize);
    }
}

File: frontend/js/views/standard_vs_itemized.js
// Standard vs itemized deduction — whichever deduction is larger, via
// /calc/standard-vs-itemized. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 }));
const signed = (n) => (n == null ? '—' : (n >= 0 ? '+$' : '−$') + Math.abs(Number(n)).toLocaleString(undefined, { maximumFractionDigits: 0 }));
const VIEW = 'standard-vs-itemized';
let lastReport = null;
let lastBody = null;

export async function renderStdVsItemized(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.itemize.h1.title">// STANDARD VS ITEMIZED</span></h1>
        <p class="muted small" data-i18n="view.itemize.hint.intro">
            Itemized deductions are SALT (state/local + property tax, capped) plus mortgage interest,
            charitable gifts, and medical costs above 7.5% of AGI. You take whichever is larger — the
            standard deduction or the itemized total. 2026 defaults: $16,100 single / $32,200 MFJ
            standard, $40,400 SALT cap (phasing down above $505k AGI). Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.itemize.h2.inputs">Your deductions</h2>
            <form id="itemize-form" class="inline-form">
                <label><span data-i18n="view.itemize.label.agi">AGI ($)</span>
                    <input type="number" step="1000" min="0" name="agi_usd" value="200000" required></label>
                <label><span data-i18n="view.itemize.label.status">Filing status</span>
                    <select name="filing_status">
                        <option value="single" data-i18n="view.itemize.status.single">Single</option>
                        <option value="married_joint" selected data-i18n="view.itemize.status.mfj">Married filing jointly</option>
                    </select></label>
                <label><span data-i18n="view.itemize.label.salt">State/local income tax ($)</span>
                    <input type="number" step="500" min="0" name="state_local_tax_usd" value="20000"></label>
                <label><span data-i18n="view.itemize.label.property">Property tax ($)</span>
                    <input type="number" step="500" min="0" name="property_tax_usd" value="10000"></label>
                <label><span data-i18n="view.itemize.label.mortgage">Mortgage interest ($)</span>
                    <input type="number" step="500" min="0" name="mortgage_interest_usd" value="18000"></label>
                <label><span data-i18n="view.itemize.label.charitable">Charitable gifts ($)</span>
                    <input type="number" step="500" min="0" name="charitable_usd" value="5000"></label>
                <label><span data-i18n="view.itemize.label.medical">Medical expenses ($)</span>
                    <input type="number" step="500" min="0" name="medical_usd" value="0"></label>
            </form>
            <div id="itemize-tools" class="ce-toolbar"></div>
        </div>
        <div id="itemize-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#itemize-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            agi_usd: Number(fd.get('agi_usd')) || 0,
            filing_status: fd.get('filing_status'),
            state_local_tax_usd: Number(fd.get('state_local_tax_usd')) || 0,
            property_tax_usd: Number(fd.get('property_tax_usd')) || 0,
            mortgage_interest_usd: Number(fd.get('mortgage_interest_usd')) || 0,
            charitable_usd: Number(fd.get('charitable_usd')) || 0,
            medical_usd: Number(fd.get('medical_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcStdVsItemized(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.itemize.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#itemize-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'standard-vs-itemized.csv' });
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['deduction_taken_usd', r.deduction_taken_usd],
        ['standard_deduction_usd', r.standard_deduction_usd],
        ['itemized_total_usd', r.itemized_total_usd],
        ['itemizing_advantage_usd', r.itemizing_advantage_usd],
        ['salt_deductible_usd', r.salt_deductible_usd],
        ['medical_deductible_usd', r.medical_deductible_usd],
    ];
}

function renderResult(mount, r) {
    const el = mount.querySelector('#itemize-result');
    const cls = r.should_itemize ? 'pos' : '';
    // Standard vs itemized deduction comparison bars (the larger wins).
    const chart = enh.svgBarChart([
        { label: 'Standard', value: r.standard_deduction_usd },
        { label: 'Itemized', value: r.itemized_total_usd },
    ]);
    const verdictKey = r.should_itemize ? 'view.itemize.verdict.itemize' : 'view.itemize.verdict.standard';
    const saltLostRow = r.salt_lost_usd > 0
        ? `<tr class="neg"><td data-i18n="view.itemize.row.salt_lost">SALT lost over cap</td><td>${money(r.salt_lost_usd)}</td></tr>`
        : '';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.itemize.h2.result">The choice</h2>
            <div class="cards">
                <div class="card ${cls}"><div class="label" data-i18n="${verdictKey}">—</div>
                    <div class="value ${cls}">${money(r.deduction_taken_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.itemize.card.standard">Standard</div>
                    <div class="value">${money(r.standard_deduction_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.itemize.card.itemized">Itemized</div>
                    <div class="value">${money(r.itemized_total_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.itemize.row.salt_paid">SALT paid</td><td>${money(r.salt_paid_usd)}</td></tr>
                    <tr><td data-i18n="view.itemize.row.salt_cap">SALT cap (after phase-out)</td><td>${money(r.salt_cap_usd)}</td></tr>
                    <tr><td data-i18n="view.itemize.row.salt_deductible">SALT deductible</td><td>${money(r.salt_deductible_usd)}</td></tr>
                    ${saltLostRow}
                    <tr><td data-i18n="view.itemize.row.medical_floor">Medical floor (7.5% AGI)</td><td>${money(r.medical_floor_usd)}</td></tr>
                    <tr><td data-i18n="view.itemize.row.medical_deductible">Medical deductible</td><td>${money(r.medical_deductible_usd)}</td></tr>
                    <tr><td data-i18n="view.itemize.row.itemized">Itemized total</td><td>${money(r.itemized_total_usd)}</td></tr>
                    <tr><td data-i18n="view.itemize.row.standard">Standard deduction</td><td>${money(r.standard_deduction_usd)}</td></tr>
                    <tr class="emph ${cls}"><td data-i18n="view.itemize.row.advantage">Itemizing advantage</td><td>${signed(r.itemizing_advantage_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}
